Blog on managing AKS managed Istio external ingress gateway with Gateway API (#4458)
Azure/AKS: blog/_posts/2024-08-02-istio-with-gateway-api.md

---
title: "Using AKS-managed Istio External Ingress Gateway with Gateway API"
description: "Learn how to use the Gateway API to manage the Istio external ingress gateway on AKS."
date: 2024-08-06
author: Paul Yu
categories: networking, operations, add-ons
---

## Introduction

> Kubernetes is great at orchestrating containers, but it can be a bit tricky to manage traffic routing. There are many options and implementations that you, as a cluster operator has probably had to deal with. We have the default Services resource that can be used to expose applications, but it is limited to routing based on layer 4 (TCP/UDP) and does not support more advanced traffic routing use cases. There's also the Ingress controller, which enabled layer 7 (HTTP) routing, and securing the North-South traffic using TLS, but it was not standardized and each vendor implementation learning a new set of resource annotations. When it comes to managing and securing East-West traffic between services, there's Service Mesh which is yet another layer of infrastructure to manage on top of Kubernetes. And we're in the same boat when it comes to resource management with each vendor having their own ways of doing things.

Reviewer: ```suggestion
Kubernetes is great at orchestrating containers, but it can be a bit tricky to manage traffic routing. There are many options and implementations that you, as a cluster operator have probably had to deal with. We have the default Service resource that can be used to expose applications, but it is limited to routing based on layer 4 (TCP/UDP) and does not support more advanced traffic routing use cases. There's also the Ingress controller, which enabled layer 7 (HTTP) routing, and securing the North-South traffic using TLS, but it was not standardized and each vendor implementation required learning a new set of resource annotations. When it comes to managing and securing East-West traffic between services, there's Service Mesh which is yet another layer of infrastructure to manage on top of Kubernetes. And we're in the same boat when it comes to resource management with each vendor having their own ways of doing things.
```

> Kubernetes is great at orchestrating containers, but it can be a bit tricky to manage traffic routing. There are many options and implementations that you, as a cluster operator has probably had to deal with. We have the default Services resource that can be used to expose applications within the cluster, but it is limited to routing based on layer 4 (TCP/UDP) and does not support more advanced traffic routing use cases. So enter the Ingress controller, which enabled layer 7 (HTTP) routing, and securing the North-South traffic using TLS, but it was not standardized and each vendor implementation learning a new set of resource annotations. When it came to managing and securing East-West traffic between services, we had to rely on a service mesh which is yet another layer of infrastructure to manage on top of Kubernetes. And we were in the same boat with configuration management with each vendor having their own custom resources ways of doing things.

Reviewer 2: Minor nit, Maybe remove "So" and just have "Enter the Ingress..."

Author: Re-worded the paragraph.

## Rise of the Gateway API

> There's no doubt that Ingress and Service Mesh have greatly improved the way we manage and secure traffic in Kubernetes, but the subtle differences between implementations has made it difficult to standardize on a single way of doing things. This is where the [Gateway API](https://gateway-api.sigs.k8s.io/) comes in. It is a Kubernetes add-on that aims to bring users an extensible, role-oriented, and protocol-aware services configuration mechanism. It is a new API that both Ingress and Service Mesh vendors can standardize on.

Reviewer: ```suggestion
There's no doubt that Ingress and service mesh have greatly improved the way we manage and secure traffic in Kubernetes, but the subtle differences between implementations has made it difficult to standardize on a single way of doing things. This is where the [Gateway API](https://gateway-api.sigs.k8s.io/) comes in. It is a Kubernetes add-on that aims to bring users an extensible, role-oriented, and protocol-aware services configuration mechanism. It is a new API that both Ingress and service mesh vendors can standardize on.
```

> The API has been designed from the ground up, taking into account the lessons learned from the Ingress and Service Mesh APIs. To me, it is a game-changer because ensures that you can implement express traffic routing configurations that were only possible via custom Ingress annotations. It also offers role-oriented configuration, which means that cluster operators can configure gateways and app developers can manage routes through the gateway. This is a big win for organizations that have a separation of duties between cluster operators and app developers. And finally, the API is portable, meaning that the API specification is supported by many implementations including [Istio](https://istio.io/latest/).

Reviewer: ```suggestion
The API has been designed from the ground up, taking into account the lessons learned from the Ingress and service mesh APIs. To me, it is a game-changer because ensures that you can express traffic routing configurations that were previously only possible via custom Ingress annotations. It also offers role-oriented configuration, which means that cluster operators can configure gateways and app developers can manage routes through the gateway. This is a big win for organizations that have a separation of duties between cluster operators and app developers. And finally, the API is portable, meaning that the API specification is supported by many implementations including [Istio](https://istio.io/latest/).
```

## Istio on AKS

If you haven't heard of Istio yet, it is a [CNCF-graduated service mesh project](https://www.cncf.io/announcements/2023/07/12/cloud-native-computing-foundation-reaffirms-istio-maturity-with-project-graduation/) that provides a uniform way to connect, manage, and secure microservices. Istio is a fully-featured service mesh that supports managing traffic flows between services, enforcing access policies, and aggregating telemetry data. It not only provides East-West traffic management and security but also North-South traffic management and security through its Ingress Gateway. We can also thank Istio for much of the inspiration behind the Gateway API - if you've worked with Istio, you'll see the similarities.

> AKS is a managed Kubernetes offering and as such many popular open-source projects are available as managed add-ons and extensions. [Istio is one of those add-ons that can be enabled on AKS](https://learn.microsoft.com/azure/aks/istio-about?WT.mc_id=containers-146848-pauyu). AKS also provides managed deployments of Istio Ingress Gateway (both internal and external) which can be used to route traffic into the cluster. In this article, I'll show you how you can experiment with the Gateway API to manage the Istio external ingress gateway on AKS.

Reviewer: ```suggestion
AKS is a managed Kubernetes offering and as such many popular open-source projects are available as managed add-ons and extensions. [Istio is one of those add-ons that can be enabled on AKS](https://learn.microsoft.com/azure/aks/istio-about?WT.mc_id=containers-146848-pauyu). AKS also provides managed deployments of Istio Ingress Gateway, which can be used to route traffic into the cluster from both internal origins within a virtual network as well as external traffic from the public internet . In this article, I'll show you how you can experiment with the Gateway API to manage the external Istio Ingress Gateway on AKS.
```

> The Gateway API is not fully supported yet supported for the Istio-based managed add-on on AKS and this should be considered an experiment for now.

## Gateway API in action with Istio on AKS

Here's a high-level overview of what we'll be doing to leverage the Gateway API to manage the Istio external ingress gateway on AKS:

1. Provision [AKS cluster with Istio enabled](https://learn.microsoft.com/azure/aks/istio-deploy-addon#install-mesh-during-cluster-creation?WT.mc_id=containers-146848-pauyu)
1. Provision [AKS managed Istio external ingress gateway](https://learn.microsoft.com/azure/aks/istio-deploy-ingress#enable-external-ingress-gateway?WT.mc_id=containers-146848-pauyu)
1. Install [Gateway API CRDs](https://gateway-api.sigs.k8s.io/guides/#installing-gateway-api)

> 1. Deploy a new [Gateway](https://gateway-api.sigs.k8s.io/concepts/api-overview/#gateway) resource in the aks-istio-ingress namespace and make sure it points to the aks-istio-ingressgateway-external service - this is the key!

Reviewer: ```suggestion
1. Deploy a new [Gateway](https://gateway-api.sigs.k8s.io/concepts/api-overview/#gateway) resource in the aks-istio-ingress namespace and configure it to program the existing AKS-managed Istio Ingress Gateway - this is the key!
```

I think this could be worded better by being slightly more general here - I'm very familiar with Gateway API but wasn't quite sure what was meant by this until reading further and seeing the specific `spec.gatewayClassName` and `spec.addresses` details required.

Let's validate that the Gateway and HTTPRoute resources were created successfully. Grab the public IP address of the Istio external ingress gateway and test the store app using a `curl` command.

```bash
INGRESS_PUBLIC_IP=$(kubectl get svc -n aks-istio-ingress aks-istio-ingressgateway-external -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
curl -IL "http://${INGRESS_PUBLIC_IP}" -H "Host: store.aks.rocks"
```

Note the `-H "Host: store.aks.rocks"` flag is used to specify the hostname in the request header. This is important because the Istio gateway is configured to route based on hostname.

Optionally, you could also add an entries to your `/etc/hosts` file.

```bash
echo "${INGRESS_PUBLIC_IP} admin.aks.rocks store.aks.rocks" | sudo tee -a /etc/hosts
```

With the host entries updated, you can test the app in your browser. Just be sure to remove the entries from your `/etc/hosts` file when you're done.

Also, don't forget to clean up the Azure resources when you're done.

```bash
az group delete --name $RG_NAME
```

## Summary

> We can see that the Gateway API is a powerful new API that provides a standardized way to configure traffic routing in Kubernetes. It is a big step forward from the Ingress and Service Mesh APIs and provides a way to express traffic routing configurations that were previously only possible via custom Ingress annotations. The Gateway API is portable and can be used with many different implementations including Istio. In this article, you learned how you can install the Gateway API CRDs into your AKS cluster and use the Gateway API to manage the AKS-managed Istio external ingress gateway.

Reviewer: ```suggestion
We can see that the Gateway API is a powerful new API that provides a standardized way to configure traffic routing in Kubernetes. It is a big step forward from the Ingress and service mesh APIs and provides a way to express traffic routing configurations that were previously only possible via custom Ingress annotations. The Gateway API is portable and can be used with many different implementations including Istio. In this article, you learned how you can install the Gateway API CRDs into your AKS cluster and use the Gateway API to manage the AKS-managed external Istio Ingress Gateway.
```
